// ng8-social-login/esm5/lib/providers/facebook-provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { OauthProvider } from './oauth-provider';
import { bindCallback, of } from 'rxjs';
import { map, mergeMap } from 'rxjs/operators';
import { Provider } from '../models/provider';
var FacebookProvider = /** @class */ (function (_super) {
    tslib_1.__extends(FacebookProvider, _super);
    function FacebookProvider(_config) {
        var _this = _super.call(this, Provider.FACEBOOK, '//connect.facebook.net/en_US/sdk.js', (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var initConfig = _this._config.initOptions;
            /** @type {?} */
            var config = initConfig.version ? initConfig : tslib_1.__assign({}, initConfig, { version: 'v2.9' });
            FB.init(config);
        })) || this;
        _this._config = _config;
        return _this;
    }
    /**
     * @return {?}
     */
    FacebookProvider.prototype.login = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return bindCallback((/**
         * @param {?} cb
         * @return {?}
         */
        function (cb) { return FB.login(cb, _this._config.loginOptions); }))().pipe(mergeMap((/**
         * @param {?} loginResponse
         * @return {?}
         */
        function (loginResponse) {
            return bindCallback((/**
             * @param {?} cb
             * @return {?}
             */
            function (cb) { return FB.api('/me?fields=name,email,picture,first_name,last_name', cb); }))().pipe(map((/**
             * @param {?} meResponse
             * @return {?}
             */
            function (meResponse) { return ((/** @type {?} */ ({
                id: meResponse.id,
                name: meResponse.name,
                email: meResponse.email,
                profileImg: "https://graph.facebook.com/" + meResponse.id + "/picture?type=normal",
                firstName: meResponse.first_name,
                lastName: meResponse.last_name,
                idToken: loginResponse.authResponse.signedRequest,
                accessToken: loginResponse.authResponse.accessToken
            }))); })));
        })));
    };
    /**
     * @return {?}
     */
    FacebookProvider.prototype.logout = /**
     * @return {?}
     */
    function () {
        FB.logout();
        return of(true);
    };
    return FacebookProvider;
}(OauthProvider));
export { FacebookProvider };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FacebookProvider.prototype._config;
}
