// ng8-social-login/bundles/ng8-social-login.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('ng8-social-login', ['exports', '@angular/core', 'rxjs', 'rxjs/operators'], factory) :
    (global = global || self, factory(global['ng8-social-login'] = {}, global.ng.core, global.rxjs, global.rxjs.operators));
}(this, function (exports, core, rxjs, operators) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var CONFIG = new core.InjectionToken('providers configuration');

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @abstract
     */
    var /**
     * @abstract
     */
    OauthProvider = /** @class */ (function () {
        function OauthProvider(provider, src, onload, async) {
            if (async === void 0) { async = true; }
            this.ID_PREFIX = 'social-login__';
            /** @type {?} */
            var id = this.ID_PREFIX + provider;
            if (document.getElementById(id)) {
                return;
            }
            /** @type {?} */
            var script = document.createElement('script');
            script.id = id;
            script.async = async;
            script.src = src;
            script.onload = onload;
            script.defer = true;
            document.head.appendChild(script);
        }
        return OauthProvider;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        OauthProvider.prototype.ID_PREFIX;
        /**
         * @abstract
         * @return {?}
         */
        OauthProvider.prototype.login = function () { };
        /**
         * @abstract
         * @return {?}
         */
        OauthProvider.prototype.logout = function () { };
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var Provider = {
        FACEBOOK: 'facebook',
        GOOGLE: 'google',
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FacebookProvider = /** @class */ (function (_super) {
        __extends(FacebookProvider, _super);
        function FacebookProvider(_config) {
            var _this = _super.call(this, Provider.FACEBOOK, '//connect.facebook.net/en_US/sdk.js', (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var initConfig = _this._config.initOptions;
                /** @type {?} */
                var config = initConfig.version ? initConfig : __assign({}, initConfig, { version: 'v2.9' });
                FB.init(config);
            })) || this;
            _this._config = _config;
            return _this;
        }
        /**
         * @return {?}
         */
        FacebookProvider.prototype.login = /**
         * @return {?}
         */
        function () {
            var _this = this;
            return rxjs.bindCallback((/**
             * @param {?} cb
             * @return {?}
             */
            function (cb) { return FB.login(cb, _this._config.loginOptions); }))().pipe(operators.mergeMap((/**
             * @param {?} loginResponse
             * @return {?}
             */
            function (loginResponse) {
                return rxjs.bindCallback((/**
                 * @param {?} cb
                 * @return {?}
                 */
                function (cb) { return FB.api('/me?fields=name,email,picture,first_name,last_name', cb); }))().pipe(operators.map((/**
                 * @param {?} meResponse
                 * @return {?}
                 */
                function (meResponse) { return ((/** @type {?} */ ({
                    id: meResponse.id,
                    name: meResponse.name,
                    email: meResponse.email,
                    profileImg: "https://graph.facebook.com/" + meResponse.id + "/picture?type=normal",
                    firstName: meResponse.first_name,
                    lastName: meResponse.last_name,
                    idToken: loginResponse.authResponse.signedRequest,
                    accessToken: loginResponse.authResponse.accessToken
                }))); })));
            })));
        };
        /**
         * @return {?}
         */
        FacebookProvider.prototype.logout = /**
         * @return {?}
         */
        function () {
            FB.logout();
            return rxjs.of(true);
        };
        return FacebookProvider;
    }(OauthProvider));
    if (false) {
        /**
         * @type {?}
         * @private
         */
        FacebookProvider.prototype._config;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var GoogleProvider = /** @class */ (function (_super) {
        __extends(GoogleProvider, _super);
        function GoogleProvider(config) {
            var _this = _super.call(this, Provider.GOOGLE, '//apis.google.com/js/platform.js', (/**
             * @return {?}
             */
            function () {
                gapi.load('auth2', (/**
                 * @return {?}
                 */
                function () {
                    /** @type {?} */
                    var conf = !config.scope ? __assign({}, config, { scope: 'email' }) : config;
                    gapi.auth2.init(conf).then((/**
                     * @param {?} auth
                     * @return {?}
                     */
                    function (auth) { return (_this._googleAuth = auth); }));
                }));
            })) || this;
            return _this;
        }
        /**
         * @return {?}
         */
        GoogleProvider.prototype.login = /**
         * @return {?}
         */
        function () {
            var _this = this;
            return rxjs.from(this._googleAuth.signIn({ prompt: 'select_account' })).pipe(operators.map((/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var profile = _this._googleAuth.currentUser.get().getBasicProfile();
                /** @type {?} */
                var accessToken = _this._googleAuth.currentUser.get().getAuthResponse(true).access_token;
                /** @type {?} */
                var idToken = _this._googleAuth.currentUser.get().getAuthResponse(true).id_token;
                return {
                    id: profile.getId(),
                    email: profile.getEmail(),
                    name: profile.getName(),
                    profileImg: profile.getImageUrl(),
                    accessToken: accessToken,
                    idToken: idToken
                };
            })));
        };
        /**
         * @return {?}
         */
        GoogleProvider.prototype.logout = /**
         * @return {?}
         */
        function () {
            var _this = this;
            return rxjs.Observable.create((/**
             * @param {?} observer
             * @return {?}
             */
            function (observer) { return observer.complete(_this._googleAuth.signOut()); }));
        };
        return GoogleProvider;
    }(OauthProvider));
    if (false) {
        /**
         * @type {?}
         * @private
         */
        GoogleProvider.prototype._googleAuth;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SocialLoginService = /** @class */ (function () {
        function SocialLoginService(config) {
            var _this = this;
            this._providers = {};
            this._providers = Object.keys(config).reduce((/**
             * @param {?} obj
             * @param {?} providerId
             * @return {?}
             */
            function (obj, providerId) {
                /** @type {?} */
                var provider = Provider[providerId.toUpperCase()];
                obj[provider] = _this.oauthProviderFactory(provider, config[providerId]);
                return obj;
            }), {});
        }
        /**
         * @param {?} provider
         * @return {?}
         */
        SocialLoginService.prototype.login = /**
         * @param {?} provider
         * @return {?}
         */
        function (provider) {
            this._providerInUse = provider;
            /** @type {?} */
            var oauthProvider = this._providers[provider];
            return oauthProvider ? oauthProvider.login() : rxjs.EMPTY;
        };
        /**
         * @return {?}
         */
        SocialLoginService.prototype.logout = /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var oauthProvider = this._providers[this._providerInUse];
            /** @type {?} */
            var ret = oauthProvider ? oauthProvider.logout() : rxjs.EMPTY;
            this._providerInUse = undefined;
            return ret;
        };
        /**
         * @return {?}
         */
        SocialLoginService.prototype.loggedIn = /**
         * @return {?}
         */
        function () {
            return this._providerInUse ? true : false;
        };
        /**
         * @private
         * @param {?} provider
         * @param {?} config
         * @return {?}
         */
        SocialLoginService.prototype.oauthProviderFactory = /**
         * @private
         * @param {?} provider
         * @param {?} config
         * @return {?}
         */
        function (provider, config) {
            switch (provider) {
                case Provider.FACEBOOK:
                    return new FacebookProvider((/** @type {?} */ (config)));
                case Provider.GOOGLE:
                    return new GoogleProvider((/** @type {?} */ (config)));
            }
        };
        SocialLoginService.decorators = [
            { type: core.Injectable }
        ];
        /** @nocollapse */
        SocialLoginService.ctorParameters = function () { return [
            { type: undefined, decorators: [{ type: core.Inject, args: [CONFIG,] }] }
        ]; };
        return SocialLoginService;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        SocialLoginService.prototype._providers;
        /**
         * @type {?}
         * @private
         */
        SocialLoginService.prototype._providerInUse;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxSocialLoginModule = /** @class */ (function () {
        function NgxSocialLoginModule() {
        }
        /**
         * @param {?} config
         * @return {?}
         */
        NgxSocialLoginModule.init = /**
         * @param {?} config
         * @return {?}
         */
        function (config) {
            return {
                ngModule: NgxSocialLoginModule,
                providers: [
                    { provide: CONFIG, useValue: config }
                ]
            };
        };
        NgxSocialLoginModule.decorators = [
            { type: core.NgModule, args: [{
                        providers: [SocialLoginService]
                    },] }
        ];
        return NgxSocialLoginModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @record
     */
    function SocialUser() { }
    if (false) {
        /** @type {?} */
        SocialUser.prototype.id;
        /** @type {?} */
        SocialUser.prototype.email;
        /** @type {?} */
        SocialUser.prototype.name;
        /** @type {?} */
        SocialUser.prototype.profileImg;
        /** @type {?|undefined} */
        SocialUser.prototype.firstName;
        /** @type {?|undefined} */
        SocialUser.prototype.lastName;
        /** @type {?} */
        SocialUser.prototype.accessToken;
        /** @type {?} */
        SocialUser.prototype.idToken;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @record
     */
    function FacebookProviderConfig() { }
    if (false) {
        /** @type {?} */
        FacebookProviderConfig.prototype.initOptions;
        /** @type {?|undefined} */
        FacebookProviderConfig.prototype.loginOptions;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @record
     */
    function GoogleProviderConfig() { }
    if (false) {
        /** @type {?} */
        GoogleProviderConfig.prototype.client_id;
        /** @type {?|undefined} */
        GoogleProviderConfig.prototype.cookie_policy;
        /** @type {?|undefined} */
        GoogleProviderConfig.prototype.scope;
        /** @type {?|undefined} */
        GoogleProviderConfig.prototype.fetch_basic_profile;
        /** @type {?|undefined} */
        GoogleProviderConfig.prototype.hosted_domain;
        /** @type {?|undefined} */
        GoogleProviderConfig.prototype.openid_realm;
        /** @type {?|undefined} */
        GoogleProviderConfig.prototype.ux_mode;
        /** @type {?|undefined} */
        GoogleProviderConfig.prototype.redirect_uri;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @record
     */
    function ProviderConfig() { }
    if (false) {
        /** @type {?|undefined} */
        ProviderConfig.prototype.google;
        /** @type {?|undefined} */
        ProviderConfig.prototype.facebook;
    }

    exports.NgxSocialLoginModule = NgxSocialLoginModule;
    exports.Provider = Provider;
    exports.SocialLoginService = SocialLoginService;
    exports.ɵa = CONFIG;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
