// ng8-social-login/esm2015/lib/providers/facebook-provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { OauthProvider } from './oauth-provider';
import { bindCallback, of } from 'rxjs';
import { map, mergeMap } from 'rxjs/operators';
import { Provider } from '../models/provider';
export class FacebookProvider extends OauthProvider {
    /**
     * @param {?} _config
     */
    constructor(_config) {
        super(Provider.FACEBOOK, '//connect.facebook.net/en_US/sdk.js', (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const initConfig = this._config.initOptions;
            /** @type {?} */
            const config = initConfig.version ? initConfig : Object.assign({}, initConfig, { version: 'v2.9' });
            FB.init(config);
        }));
        this._config = _config;
    }
    /**
     * @return {?}
     */
    login() {
        return bindCallback((/**
         * @param {?} cb
         * @return {?}
         */
        (cb) => FB.login(cb, this._config.loginOptions)))().pipe(mergeMap((/**
         * @param {?} loginResponse
         * @return {?}
         */
        (loginResponse) => bindCallback((/**
         * @param {?} cb
         * @return {?}
         */
        (cb) => FB.api('/me?fields=name,email,picture,first_name,last_name', cb)))().pipe(map((/**
         * @param {?} meResponse
         * @return {?}
         */
        (meResponse) => ((/** @type {?} */ ({
            id: meResponse.id,
            name: meResponse.name,
            email: meResponse.email,
            profileImg: `https://graph.facebook.com/${meResponse.id}/picture?type=normal`,
            firstName: meResponse.first_name,
            lastName: meResponse.last_name,
            idToken: loginResponse.authResponse.signedRequest,
            accessToken: loginResponse.authResponse.accessToken
        })))))))));
    }
    /**
     * @return {?}
     */
    logout() {
        FB.logout();
        return of(true);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    FacebookProvider.prototype._config;
}
